### PR TITLE
[lc_ctrl/otp_ctrl/doc] Documentation updates

### PR DESCRIPTION
**Pull request by msfschaffner**
Updates include:

LC:
- [x] Corrections to the programmers guide due to the multibit mutex
- [x] Update block diagram
- [x] Document signals and interfaces  
- [x] Update documentation of revised life cycle access control signals (#4504)
- [x] Add main FSM diagram
~~- [ ] Add system integration diagram~~ (will add that in a subsequent PR)

OTP:
- [x] Explicit documentation of access granularity of all OTP items
- [x] DAI FSM diagram update
- [x] Blockdiagram update

### CODE REVIEW

**Comment by cindychip** [comment]
Hey Michael, one small thing could you also update the LC spec here: https://docs.opentitan.org/hw/ip/lc_ctrl/doc/index.html#programmers-guide
Point 3 says: "Claim exclusive access to the transition interface by writing 1 to the CLAIM_TRANSITION_IF register,"
It should be writing `'hA5` right? Do you mind updating that also?

File: hw/ip/otp_ctrl/doc/_index.md [hunk 1]
```diff
@@ -945,6 +945,9 @@ For the HW_CFG and SECRET* partitions, this can be achieved as follows:
 2. [Read back]({{< relref "#readout-sequence" >}}) and verify the digest location via the DAI.
 3. Perform a full-system reset and verify that the corresponding CSRs exposing the 64bit digest have been populated ({{< regref "HW_CFG_DIGEST_0" >}}, {{< regref "SECRET0_DIGEST_0" >}}, {{< regref "SECRET1_DIGEST_0" >}} or {{< regref "SECRET2_DIGEST_0" >}}).
 
+It should be noted that locking only takes effect after a system reset since the affected partitions first have to re-sense the digest values.
+Hence, SW must ensure that no more data is written to the partition to be locked after triggering the hardware digest calculation.
```
Comment (cdgori) [inline hw/ip/otp_ctrl/doc/_index.md:949]: If I understand correctly, not doing this will likely brick the device, so maybe just clarify how essential this is?

**Review by cdgori — APPROVED** [review]
LGTM with one small suggestion

File: hw/ip/otp_ctrl/doc/_index.md (hunk 1 above)
Comment (msfschaffner) [reply hw/ip/otp_ctrl/doc/_index.md:949]: Good point, let me emphasize that.

**Comment by msfschaffner** [comment]
> Hey Michael, one small thing could you also update the LC spec here: https://docs.opentitan.org/hw/ip/lc_ctrl/doc/index.html#programmers-guide
> Point 3 says: "Claim exclusive access to the transition interface by writing 1 to the CLAIM_TRANSITION_IF register,"
> It should be writing `'hA5` right? Do you mind updating that also?

Yeah that's right, thanks for catching that. I'll amend this part.

**Comment by msfschaffner** [comment]
Ok this documentation update is mostly final now.
There is another system integration diagram for life cycle which is not quite finished yet.
I will add that in a subsequent PR.

File: hw/ip/lc_ctrl/doc/_index.md [hunk 2]
```diff
@@ -495,9 +624,9 @@ Hence the following programming sequence applies to both SW running on the devic
...
+3. Claim exclusive access to the transition interface by writing 0xA5 to the {{< regref "CLAIM_TRANSITION_IF" >}} register, and reading it back. If the value read back equals to 0xA5, the hardware mutex has successfully been claimed and SW can proceed to step 4. If the value read back equals to 0, the mutex has already been claimed by the other interface (either CSR or TAP), and SW should try claiming the mutex again.
 
-4. Write the desired target state to {{< regref "TRANSITION_TARGET" >}}. If the transition is conditional, the corresponding token should be written to {{< regref "TRANSITION_TOKEN_0" >}} as well. An optional, but recommended step is to read back and verify the values written to these registers before proceeding with step 5.
+4. Write the desired target state to {{< regref "TRANSITION_TARGET" >}}. For conditional transitions, the corresponding token has to be written to {{< regref "TRANSITION_TOKEN_0" >}}. For all unconditional transitions, the token registers have to be set to zero. An optional, but recommended step is to read back and verify the values written to these registers before proceeding with step 5.
```
Comment (cindychip) [inline hw/ip/lc_ctrl/doc/_index.md:629]: For this line: the corresponding token has to be written to {{< regref "TRANSITION_TOKEN_0" >}}
We are not just writing to TRANSITION_TOKEN_0 right? We will use TRANSITION_TOKEN_0, TRANSITION_TOKEN_1, and TRANSITION_TOKEN_2. Maybe just use TRANSITION_TOKEN_*?

Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:629]: That is true, although I found it a bit clunky to reference all registers, since they all are needed to store the full 128bit value... do you think it would be cleared to reference all registers of the multireg?

**Review by cindychip — APPROVED** [review]
LGTM with one small suggestion

File: hw/ip/lc_ctrl/doc/_index.md (hunk 2 above)
Comment (cindychip) [reply hw/ip/lc_ctrl/doc/_index.md:629]: Yeah i agree mentioning them all is a bit clunky. I was thinking about using `*` but it won't have a register link.  Yeah I am fine with what it is then :)

Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:629]: OK sounds good (I know, using the * would indicate that better, but loosing the link is also not a good option, I think...)

File: hw/ip/lc_ctrl/doc/_index.md [hunk 3]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+
+#### CREATOR_SEED_SW_RW_EN and OWNER_SEED_SW_RW_EN
+
+These signals control whether the non-volatile provisioning of life cycle related collateral can be accessed.
+The signals can only be active during DEV / PROD / PROD_END / RMA.
+During other states, it is not possible to either read or modify the collateral.
+This specifically limits the danger of rogue software images during any TEST_UNLOCKED state.
+However, as these signals only gate functional access and not DFT access, it is still possible for a malicious agent to bypass this protection by abusing scan shift/capture mechanics.
+
+While the OWNER_SEED_SW_RW_EN is statically enabled in the states shown above, the CREATOR_SEED_SW_RW_EN is only enabled if the device has not yet been personalized (i.e., the OTP partition holding the root key has not been locked down yet).
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:236]: so just one minor thing here.  The `creator_seed_sw_rw_en` is also used to control the provision of contents in flash. So this would mean that the user would need to..

1. configure otp contents
2. configure flash contents
3. now lock down otp (or swap steps 2 and 1), is that right? 

If so, we just need to make sure it is very clear that this has to be done in a specific order. 
I think this dependency is fine, we just have to be careful since it'll probably be easy for someone to do it wrong a few times and accidentally lock themselves out of configuring flash.

Let me know if I misunderstood.

File: hw/ip/lc_ctrl/doc/_index.md [hunk 4]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+#### SEED_HW_RD_EN
+
+The SEED_HW_RD_EN signal controls whether the owner and creator root keys can be accessed by hardware.
+This signal is dependent on the personalization state of the device and will only be enabled if the device has been personalized (i.e., when the OTP partition holding the root key has been locked down).
+
+#### ISO_PART_SW_RD_EN and ISO_PART_SW_WR_EN
+
+These signals control whether the isolated flash partition holding additional manufacturing details can be accessed.
+The isolated partition is both read and writable during the DEV / PROD / PROD_END / RMA states.
+In all other states it is inaccessible, except during the TEST_UNLOCKED state where the partition is write-only.
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:249]: recommend "TEST_UNLOCKED states" since there are multiple such states.

File: hw/ip/lc_ctrl/doc/_index.md [hunk 5]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+As described in more detail in the [OTP controller specification]({{< relref "hw/ip/otp_ctrl/doc/_index.md#direct-access-memory-map">}}), these digests have a non-zero value once the partition has been provisioned and write/read access has been locked.
+
+### ID State of the Device
+
+If the SECRET2_DIGEST is zero, the device is considered to have "blank" ID state, in which case the CREATOR_ROOT_KEY_* (in OTP) and CREATOR_DIV_KEY (in FLASH) can be written by software.
+All consumers of these keys are supplied with an invalid value.
+
+If the SECRET2_DIGEST has a nonzero value, the device is considered "creator personalized", and the CREATOR_ROOT_KEY and CREATOR_DIV_KEY are no longer accessible to software.
+Actual values are supplied to the consumers.
+If SECRET2_DIGEST has a nonzero value, the CREATOR_SEED_SW_RW_EN signal will be disabled in PROD, PROD_END and RMA states.
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:274]: is the last part regarding RMA states still true?

File: hw/ip/lc_ctrl/doc/_index.md [hunk 6]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+
+- *_TOKEN
+- CREATOR_ROOT_KEY*
+
+Specifically this means after OTP sensing, the above entries are unwrapped to obtain the real value.
+Similarly, during programming, they are wrapped before beginning to be written to OTP.
+
+The function used for this wrapping is the lightweight PRESENT-cipher.
+The wrapping is a one time event during controlled manufacturing, and unwrapping also cannot be supplied with arbitrary ciphertexts.
+Thus the system cannot be abused to generate a large number of traces for informational leakage, and thus a fully hardened cipher (such as AES) is not required.
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:288]: i would say `masked-AES` just in case someone wants to argue that AES can also be unhardened.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 3 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:236]: your understanding is correct - we basically consolidated `ID_STATE` with the `SECRET2_DIGEST`.

both provisioning orders you mentioned are fine (1 2 3 or 2 1 3).

where do you think would be the best place to capture this? do we have a provisioning sequence document somewhere?

File: hw/ip/lc_ctrl/doc/_index.md (hunk 4 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:249]: ah yeah good point - will fix

File: hw/ip/lc_ctrl/doc/_index.md [hunk 7]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+SEED_HW_RD_EN only becomes active if SECRET2_DIGEST has a nonzero value in DEV, PROD, PROD_END and RMA states.
+
+## Flash Collateral
+
+The flash contains both memory mapped and non-memory mapped partitions.
+As it pertains to life cycle, the flash contains three sets of important collateral.
+They are enumerated in the table below.
+Just as with OTP, the consumer and usage of each is also described.
+
+{{< snippet "hw/ip/lc_ctrl/doc/lc_ctrl_flash_collateral.md" >}}
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:317]: this is my bad for not updating. But at the moment `ownership blob` does not exist.
So you can take it out of the table.  (fine to be different PR)

File: hw/ip/lc_ctrl/doc/_index.md [hunk 8]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+### Flash Accessibility Summary and Impact of Life Cycle Signals
+
+The creator software is trusted to manage the owner partition (OWNER_DATA).
+As such, OWNER_DATA is remains accessible during DEV / PROD / PROD_END / RMA states, irrespective of the device personalization state.
+It is expected that ROM_ext during secure boot programs the protection correctly such that downstream software has appropriate permissions.
+
+The CREATOR_DATA partitions however, are further qualified based on the personalization state of the device.
+Just as with OTP, the table below enumerates accessibility of flash collateral.
+
+{{< snippet "hw/ip/lc_ctrl/doc/lc_ctrl_flash_accessibility.md" >}}
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:334]: same comment as above, the "blob" no longer exists.
I should also think about how to word this on flash side.
Maybe I should describe it more as to how it links to partitions and provide a reference here?

File: hw/ip/lc_ctrl/doc/_index.md (hunk 5 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:274]: yes I need to change this to PROD, PROD_END and DEV states

File: hw/ip/lc_ctrl/doc/_index.md (hunk 6 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:288]: ok will fix

File: hw/ip/lc_ctrl/doc/_index.md [hunk 9]
```diff
@@ -195,41 +200,220 @@ This ensures that during specific states (RAW, TEST_LOCKED, SCRAP, INVALID) it i
...
+
+#### Power Manager Interface
+
+The power manager interface is comprised of three signals overall: an initialization request (`pwr_lc_i.lc_init`), an initialization done response (`pwr_lc_o.lc_done`) and an idle indicator (`pwr_lc_o.lc_idle`).
+
+The power manager asserts `pwr_lc_i.lc_init` in order to signal to the life cycle controller that it can start initialization, and the life cycle controller signals completion of the initialization sequence by asserting `pwr_lc_o.lc_done` (the signal will remain high until reset).
+
+The idle indication signal `pwr_lc_o.lc_idle` indicates that the life cycle controller is idle.
+If this bit is 0, the life cycle controller is either not initialized or in the middle of carryingout a life cycle state transition.
+The power manager uses that indication to determine whether a power down request needs to be aborted.
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:403]: btw, off topic question, I can't remember the conclusion on otp brownout.
Is it a terrible thing if we try to shut off power when otp is actively trying to program something?

File: hw/ip/lc_ctrl/doc/_index.md (hunk 7 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:317]: ok let me remove that (I can do it here since there are a few other things that need fixing)

File: hw/ip/lc_ctrl/doc/_index.md (hunk 9 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:403]: yeah we discussed that with Nuvoton and agreed that OTP brown out does not need to be explicitly handled since we only perform state transitions in safe environments.

as for the idleness signal from LC and OTP... we can still leave these connected and handle them best-effort. WDYT?

File: hw/ip/lc_ctrl/doc/_index.md [hunk 10]
```diff
@@ -324,17 +524,51 @@ A decoded version of this counter is exposed in the {{< regref "LC_TRANSITION_CN
...
+Note that this first escalation action does not affect the life cycle state.
+
+When the second channel `esc_scrap_state` is asserted, the life cycle controller moves the life cycle state into `EscalateSt`, which behaves like a "virtual" SCRAP life cycle state.
+This transition is not permanent, and will clear upon the next power cycle.
+
+#### FSM Glitch Countermeasures
+
+The FSM has been designed to have a linear control flow that always moves in the same direction, and that always ends in a terminal state after initiating a transition request in order to make glitch attacks harder.
+A sparse FSM state encoding is employed, where each state is encoded as a 16bit word with a minimum Hamming distance of 5 w.r.t. any other state.
+The FSM state and the life cycle state vector are concurrently monitored, and if an erroneous encoding is detected, the life cycle FSM is immediately moved into the terminal `InvalidSt`, and a `lc_state_failure` alert is asserted.
```
Comment (tjaychen) [inline hw/ip/lc_ctrl/doc/_index.md:565]: does the `InvalidSt` also correspond to a virtual scrap? Or would that need to take the alert loop?

File: hw/ip/lc_ctrl/doc/_index.md (hunk 8 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:334]: ok let me remove. do you want to insert a plug pointing to the flash docs here? or vice versa?

**Review by tjaychen — APPROVED** [review]
LGTM! 
sorry about the pile of comments!

I think the point on digest != 0 locking out both otp / flash provisioning probably deserves mention in the programmer's model to ensure the user does not accidentally lock flash out.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 3 above)
Comment (tjaychen) [reply hw/ip/lc_ctrl/doc/_index.md:236]: i think that will probably come from @moidx later.
I think now, could you put a "CAUTION" section in the programmers model? :) and just point out this dependency. 
It would be good if @cindychip if aware of this also when we create top level tests.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 8 above)
Comment (tjaychen) [reply hw/ip/lc_ctrl/doc/_index.md:334]: let me work it out on flash side first and i'll come back and insert a link.
I think whatever you have is good, I just wanted to add some more language on how these partitions actually link to flash's data and info partition types.  

So that probably belongs in flash.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 9 above)
Comment (tjaychen) [reply hw/ip/lc_ctrl/doc/_index.md:403]: yeah sounds good to me. 
i do handle this in pwrmgr right now by just saying "o crap something is happening" and backs off.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 3 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:236]: ok sounds good, I'll add that :).

File: hw/ip/lc_ctrl/doc/_index.md (hunk 8 above)
Comment (tjaychen) [reply hw/ip/lc_ctrl/doc/_index.md:334]: sorry just to clarify, DON'T remove what you have!
i just meant i'll add another link later.

Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:334]: ok sounds good!

File: hw/ip/lc_ctrl/doc/_index.md (hunk 9 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:403]: yeah I think that is fine.

File: hw/ip/lc_ctrl/doc/_index.md (hunk 10 above)
Comment (msfschaffner) [reply hw/ip/lc_ctrl/doc/_index.md:565]: yes, and no this does not have to go via the alert loop.

we have a couple of states that are all equivalent to virtual scrap, with the only differences that they have a different encoding, and that we arrive in these states via different avenues.

I.e., `InvalidSt`, `PostTransitionSt` and `EscalateSt` all behave like virtual scrap, but they have different causes:
- `InvalidSt`: any error in either the life cycle state vector or the redundantly encoded FSM will immediately move the FSM into this state - no escalation required.
- `PostTransitionSt`: as the name says, every transition request eventually ends in this state
- `EscalateSt`: we will move here via the alert handler escalation path.

File: hw/ip/otp_ctrl/doc/_index.md [hunk 11]
```diff
@@ -813,6 +817,12 @@ The LIFE_CYCLE partition is the only partition that is modified in the field - b
 OTP words cannot be programmed twice, and doing so may damage the memory array.
 Hence the OTP controller performs a blank check and returns an error if a write operation is issued to an already programmed location.
 
+### Potential Side-Effects on Flash via Life Cycle
```
Comment (msfschaffner) [inline hw/ip/otp_ctrl/doc/_index.md:820]: @tjaychen @cindychip FYI: added a few words of caution, highlighting this dependence.

**Comment by msfschaffner** [comment]
Thanks a lot for the detailed review, @tjaychen.
Amended and rebased.
